src/backend/cdb/cdbpath.c

```diff
@@ -1237,25 +1269,92 @@ cdbpath_motion_for_join(PlannerInfo *root,
 			}
 		}
 	}
+
 	/*
-	 * Replicated paths shouldn't occur except UPDATE/DELETE on replicated table.
+	 * Replicated paths shouldn't occur except UPDATE/DELETE on replicated
+	 * table or modifying CTEs with DML operations on replicated table with
+	 * SELECT command above. In first case when we UPDATE/DELETE replicated
+	 * table, the flag upd_del_replicated_table is set true, and, therefore,
+	 * we need to make any other probable locus replicated in order to achieve
+	 * correct update/delete. Another case - DML over replicated table inside
+	 * CTE with SELECT above.
 	 */
-	else if (CdbPathLocus_IsReplicated(outer.locus))
+	else if (CdbPathLocus_IsReplicated(outer.locus) ||
+			 CdbPathLocus_IsReplicated(inner.locus))
 	{
-		if (root->upd_del_replicated_table > 0)
-			CdbPathLocus_MakeReplicated(&inner.move_to,
-										CdbPathLocus_NumSegments(outer.locus));
+		CdbpathMfjRel *replicated;
+		CdbpathMfjRel *other;
+
+		if (CdbPathLocus_IsReplicated(outer.locus))
+		{
+			replicated = &outer;
+			other = &inner;
+		}
 		else
 		{
-			Assert(false);
-			goto fail;
+			replicated = &inner;
+			other = &outer;
 		}
-	}
-	else if (CdbPathLocus_IsReplicated(inner.locus))
-	{
+
+		/* Common UPDATE/DELETE on replicated table */
 		if (root->upd_del_replicated_table > 0)
-			CdbPathLocus_MakeReplicated(&outer.move_to,
-										CdbPathLocus_NumSegments(inner.locus));
+			CdbPathLocus_MakeReplicated(&other->move_to,
+								CdbPathLocus_NumSegments(replicated->locus));
+		else if (hasModifyingCTESelectRecur(root))
+		{
+			if (CdbPathLocus_IsBottleneck(other->locus))
+			{
+				/*
+				 * bring locus Replicated to singleton in order to perform join
+				 * on single segment
+				 */
+				int		numsegments = CdbPathLocus_CommonSegments(replicated->locus,
+																  other->locus);
+
+				replicated->move_to = other->locus;
+				replicated->move_to.numsegments = numsegments;
+			}
+			else if (CdbPathLocus_IsPartitioned(other->locus))
+			{
+				if (!replicated->ok_to_replicate)
+				{
+					int		numsegments = CdbPathLocus_CommonSegments(replicated->locus,
+																	  other->locus);
+
+					CdbPathLocus_MakeSingleQE(&replicated->move_to, numsegments);
+					CdbPathLocus_MakeSingleQE(&other->move_to, numsegments);
+				}
+				else if (CdbPathLocus_NumSegments(replicated->locus) >=
+						 CdbPathLocus_NumSegments(other->locus))
+					return other->locus;
+
+				/*
+				 * For the case that other is a Hashed table and redistribute
+				 * clause matches other's distribute keys and , we could
+				 * redistribute Replicated to other.
```

Comment:
Do you have a test case for this branch? How to redistribute Replicated with numsegments > 1?

Reply:
```
Test case in with_clause.sql
with_dml_dr_seg1 - at 1 segment 
with_dml - hashed at 3 segments

explain (costs off)
with cte as (
    insert into with_dml_dr_seg1 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join with_dml on cte.i = with_dml.i;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice2; segments: 3)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (with_dml.i = with_dml_dr_seg1.i)
                     ->  Seq Scan on with_dml
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: with_dml_dr_seg1.i
                                 ->  Insert on with_dml_dr_seg1
                                       ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(12 rows)
```
For case Replicated with numsegments > 1 (still has to be less than Hashed numsegments to be redistributed):
```
explain (costs off)
with cte as (
    insert into t_repl_s2 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_hashed on cte.i = t_hashed.i;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice2; segments: 3)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t_repl_s2.i)
                     ->  Seq Scan on t_hashed
                     ->  Hash
                           ->  Redistribute Motion 2:3  (slice1; segments: 2)
                                 Hash Key: t_repl_s2.i
                                 ->  Insert on t_repl_s2
                                       ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(12 rows)
```

Reply:
How does `Redistribute Motion 2:3` handle duplicate rows on input?

Reply:
Probably one need to generate `Redistirbutre Motion 1:3` in this situation

Reply:
But you must execute the Insert node on both segments. And they both will send the same tuples. I think that redistribution of replicated data sets must be excepted. The only that we can do and focus them or join them as is when possible.